# 01_Data_Understanding

**Project**: Analysis of Socioeconomic Inequality in the ENEM 2023

**Author**: Luís Antonio Manfredi Sodré

**Date**: 2026

## Purpose

Explore and document every useful column in the MICRODADOS_ENEM_2023 dataset, identify inconsistencies, missing values and distributions, and record the decisions that will be applied in the notebook 02_data_cleaning.

## Dataset

- **Source:** INEP - National Institute of Educational Studies and Research
- **File:** MICRODADOS_ENEM_2023
- **Rows:** ~4 million records
- **Analyzed columns:** TP_SEXO; TP_COR_RACA; TP_ESCOLA; SG_UF_PROVA; TP_PRESENCA_CN; TP_PRESENCA_CH; TP_PRESENCA_LC; TP_PRESENCA_MT; NU_NOTA_CN; NU_NOTA_CH; NU_NOTA_LC; NU_NOTA_MT; TP_STATUS_REDACAO; NU_NOTA_REDACAO; Q006; Q022; Q024; Q025

## Column Reference

| Column | Description |
|----------|----------|
| TP_SEXO | Gender |
| TP_COR_RACA | Race/Ethnicity |
| TP_ESCOLA | School Type |
| SG_UF_PROVA | State Where the Exam Was Taken |
| TP_PRESENCA_CN | Natural Sciences Attendance |
| TP_PRESENCA_CH | Human Sciences Attendance |
| TP_PRESENCA_LC | Languages, Codes and their Technologies Attendance |
| TP_PRESENCA_MT | Mathematics Attendance |
| NU_NOTA_CN | Natural Sciences Score |
| NU_NOTA_CH | Human Sciences Score |
| NU_NOTA_LC | Languages, Codes and their Technologies Score |
| NU_NOTA_MT | Mathematics Score |
| TP_STATUS_REDACAO | Essay Status |
| NU_NOTA_REDACAO | Essay Score |
| Q006 | Monthly Family Income |
| Q022 | Household Mobile Phone Acess |
| Q024 | Household Computer Acess |
| Q025 | Household Internet Acess |

In [6]:
import duckdb
import pandas as pd

In [16]:
duckdb.sql("""
        SELECT 
            TP_SEXO,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_SEXO
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_SEXO,quantity
0,F,2411185
1,M,1522770


## TP_SEXO

Column contains no missing values or inconsistencies.

Categories: F e M.

**Action:** none.

In [17]:
duckdb.sql("""
        SELECT 
            TP_COR_RACA,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_COR_RACA
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_COR_RACA,quantity
0,3,1706798
1,1,1575848
2,2,509511
3,4,64512
4,0,52575
5,5,24711


## TP_COR_RACA

Column contains no missing values or inconsistencies.

Original numeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| 0 | Not Informed |
| 1 | White |
| 2 | Black |
| 3 | Brown (Mixed Race) |
| 4 | Asian |
| 5 | Indigenous |

**Action:** rename numeric values to descriptive labels.

In [18]:
duckdb.sql("""
        SELECT 
            TP_ESCOLA,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_ESCOLA
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_ESCOLA,quantity
0,1,2532796
1,2,1166540
2,3,234619


## TP_ESCOLA

Column contains no missing values or inconsistencies.

Original numeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| 1 | Didn't respond |
| 2 | Public |
| 3 | Private |

**Action:** rename numeric values to descriptive labels.

In [19]:
duckdb.sql("""
        SELECT 
            SG_UF_PROVA,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY SG_UF_PROVA
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,SG_UF_PROVA,quantity
0,SP,590767
1,MG,358575
2,BA,324268
3,RJ,282296
4,CE,241960
5,PA,229162
6,PE,218859
7,PR,166506
8,MA,165756
9,RS,159919


## SG_UF_PROVA

Column contains no missing values or inconsistencies.

**Action:** none.

In [20]:
duckdb.sql("""
        SELECT 
            TP_PRESENCA_CN,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_PRESENCA_CN
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_PRESENCA_CN,quantity
0,1,2692427
1,0,1239316
2,2,2212


## TP_PRESENCA_CN

Column contains no missing values or inconsistencies.

Original numeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| 0 | Absent |
| 1 | Present |
| 2 | Eliminated |

**Action:** rename numeric values to descritive labels.

In [21]:
duckdb.sql("""
        SELECT 
            TP_PRESENCA_CH,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_PRESENCA_CH
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_PRESENCA_CH,quantity
0,1,2822643
1,0,1106714
2,2,4598


## TP_PRESENCA_CH

Column contains no missing values or inconsistencies.

Original numeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| 0 | Absent |
| 1 | Present |
| 2 | Eliminated |

**Action:** rename numeric values to descritive labels.

In [22]:
duckdb.sql("""
        SELECT 
            TP_PRESENCA_LC,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_PRESENCA_LC
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_PRESENCA_LC,quantity
0,1,2822643
1,0,1106714
2,2,4598


## TP_PRESENCA_LC

Column contains no missing values or inconsistencies.

Original numeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| 0 | Absent |
| 1 | Present |
| 2 | Eliminated |

**Action:** rename numeric values to descritive labels.

In [23]:
duckdb.sql("""
        SELECT 
            TP_PRESENCA_MT,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_PRESENCA_MT
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_PRESENCA_MT,quantity
0,1,2692427
1,0,1239316
2,2,2212


## TP_PRESENCA_MT

Column contains no missing values or inconsistencies.

Original numeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| 0 | Absent |
| 1 | Present |
| 2 | Eliminated |

**Action:** rename numeric values to descritive labels.

In [10]:
duckdb.sql("""
        SELECT 
            MIN(NU_NOTA_CN) AS  Minimum,
            MAX(NU_NOTA_CN) AS Maximum,
            AVG(NU_NOTA_CN) AS Average,
            MEDIAN(NU_NOTA_CN) AS Median,
            COUNT(*) AS Total,
            COUNT(NU_NOTA_CN) AS Filled,
            COUNT(*) - COUNT(NU_NOTA_CN) as Null
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
    """).df()

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Minimum,Maximum,Average,Median,Total,Filled,Null
0,0.0,868.4,495.751547,493.9,3933955,2692427,1241528


## NU_NOTA_CN

Natural Sciences score. Scale ranges from 0 to 1000.

Missing values represent exam absence and are therefore expected.

| Metric | Value |
|---|---|
| Minimum | 0.0 |
| Maximum | 868.4 |
| Average | 495.8 |
| Median | 493.9 |
| Total Records | 3933955 |
| Non-Null Values | 2692427 |
| Missing Values | 1241528 |

**Action:** none. Missing values will be retained because they represent legitimate exam absences.

In [11]:
duckdb.sql("""
        SELECT 
            MIN(NU_NOTA_CH) AS  Minimum,
            MAX(NU_NOTA_CH) AS Maximum,
            AVG(NU_NOTA_CH) AS Average,
            MEDIAN(NU_NOTA_CH) AS Median,
            COUNT(*) AS Total,
            COUNT(NU_NOTA_CH) AS Filled,
            COUNT(*) - COUNT(NU_NOTA_CH) as Null
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Minimum,Maximum,Average,Median,Total,Filled,Null
0,0.0,823.0,523.354206,530.4,3933955,2822643,1111312


## NU_NOTA_CH

Human Sciences score. Scale ranges from 0 to 1000.

Missing values represent exam absence and are therefore expected.

| Metric | Value |
|---|---|
| Minimum | 0.0 |
| Maximum | 823.0 |
| Average | 523.4 |
| Median | 530.4 |
| Total Records | 3933955 |
| Non-Null Values | 2822643 |
| Missing Values | 1111312 |

**Action:** none. Missing values will be retained because they represent legitimate exam absences.

In [12]:
duckdb.sql("""
        SELECT 
            MIN(NU_NOTA_LC) AS  Minimum,
            MAX(NU_NOTA_LC) AS Maximum,
            AVG(NU_NOTA_LC) AS Average,
            MEDIAN(NU_NOTA_LC) AS Median,
            COUNT(*) AS Total,
            COUNT(NU_NOTA_LC) AS Filled,
            COUNT(*) - COUNT(NU_NOTA_LC) as Null
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Minimum,Maximum,Average,Median,Total,Filled,Null
0,0.0,820.8,518.146637,523.1,3933955,2822643,1111312


## NU_NOTA_LC

Languages, Codes and their Technologies score. Scale ranges from 0 to 1000.

Missing values represent exam absence and are therefore expected.

| Metric | Value |
|---|---|
| Minimum | 0.0 |
| Maximum | 820.8 |
| Average | 518.1 |
| Median | 523.1 |
| Total Records | 3933955 |
| Non-Null Values | 2822643 |
| Missing Values | 1111312 |

**Action:** none. Missing values will be retained because they represent legitimate exam absences.

In [13]:
duckdb.sql("""
        SELECT 
            MIN(NU_NOTA_MT) AS  Minimum,
            MAX(NU_NOTA_MT) AS Maximum,
            AVG(NU_NOTA_MT) AS Average,
            MEDIAN(NU_NOTA_MT) AS Median,
            COUNT(*) AS Total,
            COUNT(NU_NOTA_MT) AS Filled,
            COUNT(*) - COUNT(NU_NOTA_MT) as Null
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Minimum,Maximum,Average,Median,Total,Filled,Null
0,0.0,958.6,533.8357,523.6,3933955,2692427,1241528


## NU_NOTA_MT

Mathematics score. Scale ranges from 0 to 1000.

Missing values represent exam absence and are therefore expected.

| Metric | Value |
|---|---|
| Minimum | 0.0 |
| Maximum | 958.6 |
| Average | 533.8 |
| Median | 523.6 |
| Total Records | 3933955 |
| Non-Null Values | 2692427 |
| Missing Values | 1241528 |

**Action:** none. Missing values will be retained because they represent legitimate exam absences.

In [24]:
duckdb.sql("""
        SELECT 
            TP_STATUS_REDACAO,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY TP_STATUS_REDACAO
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,TP_STATUS_REDACAO,quantity
0,1,2704814
1,<NA>,1111312
2,4,55403
3,6,24466
4,3,22505
5,8,9091
6,9,2443
7,2,2016
8,7,1905


## TP_STATUS_REDACAO

Column contains missing values, which will be ignored for category analysis.

Original numeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| 1 | No Issues |
| 2 | Annuled |
| 3 | Copy of Motivational Text |
| 4 | Blank Essay |
| 6 | Off-Topic |
| 7 | Invalid Text Type |
| 8 | Insufficient Text |
| 9 | Disconnected Contetent |

**Action:** rename numeric values to descriptive labels.

In [14]:
duckdb.sql("""
        SELECT 
            MIN(NU_NOTA_REDACAO) AS  Minimum,
            MAX(NU_NOTA_REDACAO) AS Maximum,
            AVG(NU_NOTA_REDACAO) AS Average,
            MEDIAN(NU_NOTA_REDACAO) AS Median,
            COUNT(*) AS Total,
            COUNT(NU_NOTA_REDACAO) AS Filled,
            COUNT(*) - COUNT(NU_NOTA_REDACAO) as Null
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Minimum,Maximum,Average,Median,Total,Filled,Null
0,0,1000,617.798482,620.0,3933955,2822643,1111312


## NU_NOTA_REDACAO

Essay score. Scale ranges from 0 to 1000.

Missing values represent exam absence and are therefore expected.

| Metric | Value |
|---|---|
| Minimum | 0 |
| Maximum | 1000 |
| Average | 617.8 |
| Median | 620.0 |
| Total Records | 3933955 |
| Non-Null Values | 2822643 |
| Missing Values | 1111312 |

**Action:** none. Missing values will be retained because they represent legitimate exam absences.

In [25]:
duckdb.sql("""
        SELECT 
            Q006,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY Q006
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Q006,quantity
0,B,1245271
1,C,650942
2,D,437366
3,E,293994
4,A,268053
5,G,261327
6,F,171344
7,H,139279
8,I,85970
9,J,75179


## Q006 - What is your family's monthly income? (Include your income and that of your family members)

Column contains no missing values or inconsistencies.

Original alphanumeric categories and their meanings according to the INEP data dictionary:

| Original value | Category (BRL) |
|---|---|
| A | No Income |
| B | Up to 1.320,00 |
| C | From 1.320,01 to 1.980,00. |
| D | From 1.980,01 to 2.640,00. |
| E | From 2.640,01 to 3.300,00. |
| F | From 3.300,01 to 3.960,00. |
| G | From 3.960,01 to 5.280,00. |
| H | From 5.280,01 to 6.600,00. |
| I | From 6.600,01 to 7.920,00. |
| J | From 7.920,01 to 9.240,00. |
| K | From 9.240,01 to 10.560,00. |
| L | From 10.560,01 to 11.880,00. |
| M | From 11.880,01 to 13.200,00. |
| N | From 13.200,01 to 15.840,00. |
| O | From 15.840,01 to 19.800,00. |
| P | From 19.800,01 to 26.400,00. |
| Q | Above 26.400,00. |

**Action:** rename alphanumeric values to descriptive labels.

In [26]:
duckdb.sql("""
        SELECT 
            Q022,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY Q022
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Q022,quantity
0,D,1138368
1,C,1087753
2,E,841389
3,B,764861
4,A,101584


## Q022 - Does your household have a mobile phone?

Column contains no missing values or inconsistencies.

Original alphanumeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| A | No. |
| B | Yes, one. |
| C | Yes, two. |
| D | Yes, three. |
| E | Yes, four or more. |

**Action:** rename alphanumeric values to descriptive labels.

In [27]:
duckdb.sql("""
        SELECT 
            Q024,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY Q024
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Q024,quantity
0,A,2041637
1,B,1362634
2,C,341120
3,D,126296
4,E,62268


## Q024 - Does your household have a computer?

Column contains no missing values or inconsistencies.

Original alphanumeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| A | No. |
| B | Yes, one. |
| C | Yes, two. |
| D | Yes, three. |
| E | Yes, four or more. |

**Action:** rename alphanumeric values to descriptive labels.

In [28]:
duckdb.sql("""
        SELECT 
            Q025,
            COUNT(*) AS quantity
        FROM read_csv(
            '../data/raw/MICRODADOS_ENEM_2023.csv',
            delim=';',
            encoding='latin-1'
        )
        GROUP BY Q025
        ORDER BY quantity DESC
    """).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Q025,quantity
0,B,3558451
1,A,375504


## Q025 - Does your household have Internet?

Column contains no missing values or inconsistencies.

Original alphanumeric categories and their meanings according to the INEP data dictionary:

| Original value | Category |
|---|---|
| A | No. |
| B | Yes. |

**Action:** rename alphanumeric values to descriptive labels.